scheds/rust/scx_mitosis/src/bpf/mitosis.bpf.c

```diff
@@ -0,0 +1,932 @@
+	__type(key, u32);
+	__type(value, struct cpu_ctx);
+	__uint(max_entries, 1);
+} cpu_ctxs SEC(".maps");
+
+static inline struct cpu_ctx *lookup_cpu_ctx(int cpu)
+{
+	struct cpu_ctx *cctx;
+	u32 zero = 0;
+
+	if (cpu < 0)
+		cctx = bpf_map_lookup_elem(&cpu_ctxs, &zero);
+	else
+		cctx = bpf_map_lookup_percpu_elem(&cpu_ctxs, &zero, cpu);
+
+	if (!cctx) {
+		scx_bpf_error("no cpu_ctx for cpu %d", cpu);
+		return NULL;
+	}
+
+	return cctx;
+}
+
+struct cell {
+	u64 vtime_now;
+	u32 dsq;
+	// The following field is populated from userspace to indicate
+	// which cpus the cell should belong to.
+	unsigned char cpus[MAX_CPUS_U8];
+};
+
+struct cell cells[MAX_CELLS];
+
+static inline struct cell *lookup_cell(int idx)
+{
+	if (idx < 0 || idx >= MAX_CELLS) {
+		scx_bpf_error("invalid cell %d", idx);
+		return NULL;
+	}
+	return &cells[idx];
+}
+
+/*
+ * Store the cpumask for each cell (owned by BPF logic)
+ */
+struct cell_cpumask_wrapper {
+	struct bpf_cpumask __kptr *cpumask;
+};
+
+struct {
+	__uint(type, BPF_MAP_TYPE_ARRAY);
+	__type(key, u32);
+	__type(value, struct cell_cpumask_wrapper);
+	__uint(max_entries, MAX_CELLS);
+	__uint(map_flags, 0);
+} cell_cpumasks SEC(".maps");
+
+static inline struct cpumask *lookup_cell_cpumask(int idx)
+{
+	struct cell_cpumask_wrapper *cpumaskw;
+
+	if (!(cpumaskw = bpf_map_lookup_elem(&cell_cpumasks, &idx))) {
+		scx_bpf_error("no cell cpumask");
+		return NULL;
+	}
+
+	return (struct cpumask *)cpumaskw->cpumask;
+}
+
+/* css iteration is experimental, forward declare the necessary functions */
+struct bpf_iter_css;
+struct cgroup_subsys_state;
+extern int bpf_iter_css_new(struct bpf_iter_css *it,
+			    struct cgroup_subsys_state *start,
+			    unsigned int flags) __weak __ksym;
+extern struct cgroup_subsys_state *
+bpf_iter_css_next(struct bpf_iter_css *it) __weak __ksym;
+extern void bpf_iter_css_destroy(struct bpf_iter_css *it) __weak __ksym;
+
+/*
+* Along with a user_global_seq bump, indicates that cgroup->cell assignment
+* changed
+*/
+volatile bool update_cell_assignment;
+bool draining;
+
+/*
+ * This is the main driver for reconfiguration. It only runs on CPU 0
+ * To avoid
+ */
+SEC("fentry/scheduler_tick")
+int scheduler_tick_fentry(const void *ctx)
+{
+	int cell_idx, cpu_idx;
+	struct cpu_ctx *cpu_ctx;
+	struct bpf_cpumask *cpumask;
+	struct cell_cpumask_wrapper *cell_cpumaskw;
+	struct cgroup_subsys_state *root_css, *pos;
+	struct cgroup *root_cgrp;
+
+	if (bpf_get_smp_processor_id() != 0)
+		return 0;
+
+	/*
+	 * To handle races where tasks are assigned to cells that are getting
+	 * removed, we ensure cpus dispatch from their previous cell for an entire
+	 * scheduler tick. This is a crude way of mimicing RCU synchronization.
+	 */
+	if (draining) {
+		bpf_for(cpu_idx, 0, nr_possible_cpus)
+		{
+			if (!(cpu_ctx = lookup_cpu_ctx(cpu_idx)))
+				return 0;
+
+			cpu_ctx->prev_cell = cpu_ctx->cell;
+		}
+		barrier();
+		draining = false;
+	}
+
+	if (global_seq == user_global_seq)
+		return 0;
```

Comment:
Something else to consider is doing all of these updates synchronously from inside of a `BPF_PROG_RUN` program, so you don't have to worry about user vs. global sequence numbers. We're not doing this anywhere yet because it only works on newer kernels, but it is available and potentially cleaner.

Reply:
Yeah, I'd like to do something like that. Right now I'm reliant on the tick behavior to mimic RCU, but if we had some way to actually do RCU + BPF then I'd be capable of using BPF_PROG_RUN.

Out of curiosity, is this available through libbpf_rs?

Reply:
Ah, makes sense r.e. RCU. Yes, it was made available in libbpf with this PR: https://github.com/libbpf/libbpf-rs/pull/710/files.